filament/backend/src/vulkan/VulkanPipelineCache.h

```diff
@@ -208,32 +205,63 @@ class VulkanPipelineCache : public CommandBufferObserver {
     // PIPELINE CACHE KEY
     // ------------------
 
+    // Equivalent to VkVertexInputAttributeDescription but half as big.
+    struct VertexInputAttributeDescription {
+        VertexInputAttributeDescription& operator=(const VkVertexInputAttributeDescription& that) {
+            assert_invariant(that.location <= 0xffu);
+            assert_invariant(that.binding <= 0xffu);
+            assert_invariant(uint32_t(that.format) <= 0xffffu);
+            location = that.location;
+            binding = that.binding;
+            format = that.format;
+            offset = that.offset;
+            return *this;
+        }
+        operator VkVertexInputAttributeDescription() const {
+            return { location, binding, VkFormat(format), offset };
+        }
+        uint8_t     location;
+        uint8_t     binding;
+        uint16_t    format;
+        uint32_t    offset;
+    };
+
+    // Equivalent to VkVertexInputBindingDescription but not as big.
+    struct VertexInputBindingDescription {
+        VertexInputBindingDescription& operator=(const VkVertexInputBindingDescription& that) {
+            assert_invariant(that.binding <= 0xffffu);
+            binding = that.binding;
+            stride = that.stride;
+            inputRate = that.inputRate;
+            return *this;
+        }
+        operator VkVertexInputBindingDescription() const {
+            return { binding, stride, (VkVertexInputRate) inputRate };
+        }
+        uint16_t    binding;
+        uint16_t    inputRate;
+        uint32_t    stride;
+    };
+
     // The pipeline key is a POD that represents all currently bound states that form the immutable
     // VkPipeline object.
     struct PipelineKey {
-        VkShaderModule shaders[SHADER_MODULE_COUNT]; // 16 bytes
-        RasterState rasterState;                     // 124 bytes
-        VkPrimitiveTopology topology;                // 4 bytes
-        VkRenderPass renderPass;                     // 8 bytes
-        uint16_t subpassIndex;                       // 2 bytes
-        uint16_t padding0;                           // 2 bytes
-        VkVertexInputAttributeDescription vertexAttributes[VERTEX_ATTRIBUTE_COUNT]; // 256 bytes
-        VkVertexInputBindingDescription vertexBuffers[VERTEX_ATTRIBUTE_COUNT];      // 192 bytes
-        uint32_t padding1;                                                          // 4 bytes
+        VkShaderModule shaders[SHADER_MODULE_COUNT]; // 0
+        VkRenderPass renderPass;                     // 16
+        uint16_t topology : 16;                      // 24
```

Comment:
i s the bitfield needed here, since it's a `uint16_t`?

Reply:
I think it would be very useful to have the size of each field in comment as well